sematic/caching/caching.py

```diff
@@ -0,0 +1,106 @@
+# Standard Library
+import hashlib
+import json
+import logging
+from typing import Any, Callable, Optional, Union
+
+# Sematic
+from sematic.abstract_future import AbstractFuture
+from sematic.types.serialization import value_to_json_encodable
+
+CacheNamespaceCallable = Callable[[AbstractFuture], str]
+CacheNamespace = Optional[Union[str, CacheNamespaceCallable]]
+
+logger = logging.getLogger(__name__)
+
+
+def get_future_cache_key(
+    cache_namespace: CacheNamespace, future: AbstractFuture
+) -> str:
+    """
+    Generates a cache key that can be used to uniquely identify the output value of a
+    deterministic function.
+
+    Parameters
+    ----------
+    cache_namespace: CacheNamespace
+        A string or a `Callable` which returns a string, which is used as the cache key
+        namespace.
+    future: AbstractFuture
+        The future for which to calculate the cache key.
+
+    Returns
+    -------
+    A cache key string that can be used to uniquely identify the output value of a
+    deterministic function.
+    """
+    # this may raise, so do it early
+    cache_namespace = _resolve_namespace(cache_namespace=cache_namespace, future=future)
+
+    func_fqpn = future.calculator.get_func_fqpn()  # type: ignore # noqa: ignore
+    output_type_repr = repr(future.calculator.output_type)
```

Comment:
Should we use a more stable repr here? See `db.models.factories._get_value_sha1_digest` where we use the type serialization and sort keys.